activesupport/lib/active_support/core_ext/module/attribute_accessors.rb

```diff
@@ -203,4 +203,57 @@ def mattr_accessor(*syms, instance_reader: true, instance_writer: true, instance
     mattr_writer(*syms, instance_writer: instance_writer, instance_accessor: instance_accessor, default: default, location: location)
   end
   alias :cattr_accessor :mattr_accessor
+
+  def singleton_attr_reader(*syms, instance_reader: true, instance_accessor: true, default: nil, location: nil)
+    raise TypeError, "singleton attributes should be defined directly on class, not singleton" if singleton_class?
+    location ||= caller_locations(1, 1).first
+
+    syms.each do |sym|
+      singleton_class.attr_reader(sym)
+
+      if instance_reader && instance_accessor
+        define_singleton_attr_instance_reader(sym)
+      end
+
+      sym_default_value = (block_given? && default.nil?) ? yield : default
+      instance_variable_set("@#{sym}", sym_default_value) unless sym_default_value.nil? && class_variable_defined?("@#{sym}")
+    end
+  end
+
+  def singleton_attr_writer(*syms, instance_writer: true, instance_accessor: true, default: nil, location: nil)
+    raise TypeError, "singleton attributes should be defined directly on class, not singleton" if singleton_class?
+    location ||= caller_locations(1, 1).first
+
+    syms.each do |sym|
+      singleton_class.attr_writer(sym)
+
+      if instance_writer && instance_accessor
+        define_singleton_attr_instance_writer(:"#{sym}=")
+      end
+
+      sym_default_value = (block_given? && default.nil?) ? yield : default
+      instance_variable_set("@#{sym}", sym_default_value) unless sym_default_value.nil? && class_variable_defined?("@#{sym}")
+    end
+  end
+
+  def singleton_attr_accessor(*syms, instance_reader: true, instance_writer: true, instance_accessor: true, default: nil, &blk)
+    location = caller_locations(1, 1).first
+    singleton_attr_reader(*syms, instance_reader: instance_reader, instance_accessor: instance_accessor, default: default, location: location, &blk)
+    singleton_attr_writer(*syms, instance_writer: instance_writer, instance_accessor: instance_accessor, default: default, location: location)
+  end
+
+  private
+    def define_singleton_attr_instance_reader(sym)
+      mod = self
+      define_method(sym) do
+        mod.send(sym)
+      end
+    end
+
+    def define_singleton_attr_instance_writer(sym)
+      mod = self
+      define_method(sym) do |value|
+        mod.send(sym, value)
+      end
```

Comment:
Personally, I really dislike the instance writer being enabled by default. I think it's surprising and dangerous that setting a value on an instance changes it for the class (and therefore all instances). IMO it would be great if we didn't have this (or at least wasn't on by default) for this new type of accessor declaration.

That said, that's how `cattr_writer`/`class_attribute` work, so maybe it's best to keep this the same for consistency and ease of switching 🤷‍♂️.

Reply:
> I really dislike the instance writer being enabled by default.

Me too. I did so only to keep the same interface than `cattr_accessor`, and I do remember a few cases on totally unintended instance accessors because of it in Rails.

Reply:
Let's consider this out of scope 🙂. Maybe we can address that in a different PR (maybe a global config to change the default behaviour).

Reply:
Agreed.